api/renter_test.go

```diff
@@ -30,6 +30,100 @@ func createRandFile(path string, size int) error {
 	return ioutil.WriteFile(path, data, 0600)
 }
 
+// TestRenterAsyncDownload tests that the /renter/downloadasync route works
+// correctly.
+func TestRenterAsyncDownload(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	t.Parallel()
+
+	st, err := createServerTester("TestRenterAsyncDownload")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer st.server.Close()
+
+	// Announce the host and start accepting contracts.
+	err = st.announceHost()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = st.acceptContracts()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = st.setHostStorage()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Set an allowance for the renter, allowing a contract to be formed.
+	allowanceValues := url.Values{}
+	testFunds := "10000000000000000000000000000" // 10k SC
+	testPeriod := "10"
+	allowanceValues.Set("funds", testFunds)
+	allowanceValues.Set("period", testPeriod)
+	err = st.stdPostAPI("/renter", allowanceValues)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a file.
+	path := filepath.Join(build.SiaTestingDir, "api", "TestRenterAsyncDownload", "test.dat")
+	err = createRandFile(path, 1024)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Upload to host.
+	uploadValues := url.Values{}
+	uploadValues.Set("source", path)
+	uploadValues.Set("renew", "true")
+	err = st.stdPostAPI("/renter/upload/test.dat", uploadValues)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Only one piece will be uploaded (10% at current redundancy).
+	var rf RenterFiles
+	for i := 0; i < 200 && (len(rf.Files) != 1 || rf.Files[0].UploadProgress < 10); i++ {
+		st.getAPI("/renter/files", &rf)
+		time.Sleep(100 * time.Millisecond)
+	}
+	if len(rf.Files) != 1 || rf.Files[0].UploadProgress < 10 {
+		t.Fatal("the uploading is not succeeding for some reason:", rf.Files[0])
+	}
+
+	// download the file asynchronously
+	downpath := filepath.Join(st.dir, "asyncdown.dat")
+	err = st.getAPI("/renter/downloadasync/test.dat?destination="+downpath, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
```

Comment:
download should be big enough that you can run a check here to verify that the download has not yet completed, before you spin to check that the download has completed. This test is not great atm because you could swap in the synchronous download call above and the test would still pass.